### PR TITLE
fix writepng and premultiplied alpha

### PR DESCRIPTION
**Pull request by sinkingsugar**
as well as flag premul svg image output

### CODE REVIEW

File: src/core/blocks/serialization.cpp [hunk 1]
```diff
@@ -385,9 +387,75 @@ struct WritePNG : public FileBase {
...
+            uint8_t g =
+                input.payload.imageValue.data[(ih * lineSize) + (iw * 4) + 1];
+            _scratch[(ih * lineSize) + (iw * 4) + 1] =
+                uint8_t(float(g) / fa + 0.5);
+            uint8_t b =
+                input.payload.imageValue.data[(ih * lineSize) + (iw * 4) + 2];
+            _scratch[(ih * lineSize) + (iw * 4) + 2] =
+                uint8_t(float(b) / fa + 0.5);
+          } else if (pixsize == 2) {
+            uint16_t *source =
```
Comment (sinkingsugar) [inline src/core/blocks/serialization.cpp:417]: Notice, no coverage yet

File: src/core/blocks/serialization.cpp [hunk 2]
```diff
@@ -385,9 +387,75 @@ struct WritePNG : public FileBase {
...
+            dest[(ih * lineSize) + (iw * 4) + 0] =
+                uint16_t(float(r) / fa + 0.5);
+            uint16_t g = source[(ih * lineSize) + (iw * 4) + 1];
+            dest[(ih * lineSize) + (iw * 4) + 1] =
+                uint16_t(float(g) / fa + 0.5);
+            uint16_t b = source[(ih * lineSize) + (iw * 4) + 2];
+            dest[(ih * lineSize) + (iw * 4) + 2] =
+                uint16_t(float(b) / fa + 0.5);
+          } else if (pixsize == 4) {
+            float *source =
```
Comment (sinkingsugar) [inline src/core/blocks/serialization.cpp:433]: Notice, no coverage yet

File: src/core/blocks/serialization.cpp [hunk 3]
```diff
@@ -385,9 +387,75 @@ struct WritePNG : public FileBase {
...
-      throw ActivationError("Failed to write PNG file.");
+
+    // demultiply alpha if needed, limited to 4 channels
+    if (c == 4 &&
+        (input.payload.imageValue.flags & CBIMAGE_FLAGS_PREMULTIPLIED_ALPHA) ==
+            CBIMAGE_FLAGS_PREMULTIPLIED_ALPHA) {
+      _scratch.resize(w * h * 4 * pixsize);
+      int lineSize = w * 4;
+      for (int ih = 0; ih < h; ih++) {
+        for (int iw = 0; iw < w; iw++) {
```
Comment (Kryptos-FR) [inline src/core/blocks/serialization.cpp:398]: Is it done in two nested loops for better readability or is there another reason?

File: src/core/blocks/serialization.cpp [hunk 4]
```diff
@@ -385,9 +387,75 @@ struct WritePNG : public FileBase {
...
+          if (pixsize == 1) {
+            uint8_t a =
+                input.payload.imageValue.data[(ih * lineSize) + (iw * 4) + 3];
+            float fa = float(a) / 255.0f;
+            _scratch[(ih * lineSize) + (iw * 4) + 3] = a;
+            uint8_t r =
+                input.payload.imageValue.data[(ih * lineSize) + (iw * 4) + 0];
+            _scratch[(ih * lineSize) + (iw * 4) + 0] =
+                uint8_t(float(r) / fa + 0.5);
```
Comment (Kryptos-FR) [inline src/core/blocks/serialization.cpp:407]: Might save a few CPU cycles to do the index calc only once:
```cpp
auto index = (ih * lineSize) + (iw * 4) + 0;
uint8_t r = input.payload.imageValue.data[index];
 _scratch[index] = uint8_t(float(r) / fa + 0.5);
```

File: src/core/blocks/serialization.cpp [hunk 5]
```diff
@@ -361,6 +361,8 @@ struct LoadImage : public FileBase {
 };
 
 struct WritePNG : public FileBase {
+  std::vector<uint8_t> _scratch;
```
Comment (Kryptos-FR) [inline src/core/blocks/serialization.cpp:364]: Is it necessary to have a field used as buffer? Can't it be a temporary variable?

**Review by Kryptos-FR — COMMENTED** [review]
A few remarks, but algo looks good otherwise.

File: src/core/blocks/serialization.cpp (hunk 3 above)
Comment (sinkingsugar) [reply src/core/blocks/serialization.cpp:398]: Just simplicity, I think any half decent compiler will optimize properly, but not to the extent of packing into 32 bit int.
The thing is that this block is slow in any case, we don't really need to optimize them yet.
Using (Await (WritePNG)) is advised (will run on a worker thread)

Comment (sinkingsugar) [reply src/core/blocks/serialization.cpp:398]: If I pack in 32 bits then 16bit and 32bit should be done with vectors :D - let's just add comments about it for really when there is time to waste

File: src/core/blocks/serialization.cpp (hunk 4 above)
Comment (sinkingsugar) [reply src/core/blocks/serialization.cpp:407]: I think any half decent compiler will optimize those really but for readability let's do that

File: src/core/blocks/serialization.cpp (hunk 5 above)
Comment (sinkingsugar) [reply src/core/blocks/serialization.cpp:364]: Yeah caching, if for some reason this block is called many times there will be no allocations at all (if size <=)

Comment (sinkingsugar) [reply src/core/blocks/serialization.cpp:364]: Btw this is often done all over the codebase exactly for this reason.

**Review by Kryptos-FR — APPROVED** [review]

**Pull request merged**